linera-core/src/client/mod.rs

```diff
@@ -422,6 +423,53 @@ where
     }
 }
 
+// Tracks which chain_id<>block_height pairs have been successfully processed.
+struct MultichainTracker {
+    // Starting tracker index.
+    tracker: u64,
+    // Log of received certificates by the remote node.
+    remote_node_received_log: Vec<ChainAndHeight>,
+    // Track which certificates have been successfully processed.
+    // We will use that information later to update the tracker.
+    successfuly_processed: BTreeMap<ChainId, BTreeSet<BlockHeight>>,
+}
+
+impl MultichainTracker {
+    fn new(init: u64, log: Vec<ChainAndHeight>) -> Self {
+        Self {
+            tracker: init,
+            remote_node_received_log: log,
+            successfuly_processed: BTreeMap::new(),
+        }
+    }
+
+    fn push(&mut self, chain_id: ChainId, height: BlockHeight) {
+        self.successfuly_processed
+            .entry(chain_id)
+            .or_default()
+            .insert(height);
+    }
+
+    // Returns an index of the last processed certificate in the original remote_node_received_log.
+    //
+    // Goes through the original `remote_node_received_log` and checks which of the certificates have been received.
+    // Upon finding the first entry that has not been processed, it stops and returns the index of the last processed certificate.
+    fn finalize(&mut self) -> u64 {
+        for r in &self.remote_node_received_log {
+            if let Some(processed) = self.successfuly_processed.get(&r.chain_id) {
+                if processed.contains(&r.height) {
+                    self.tracker += 1;
+                } else {
+                    break;
+                }
+            } else {
+                break;
+            }
```

Comment:
(Maybe more concise with `is_some_and`; then we need only one `break`.)

```suggestion
            if self.successfuly_processed
                   .get(&r.chain_id)
                   .is_some_and(|processed| processed.contains(&r.height))
            {
                self.tracker += 1;
            } else {
                break;
            }
```

Reply:
`is_some_and` 😵